webbpsf/wfirst.py

```diff
@@ -323,5 +324,225 @@ def _validateConfig(self, **kwargs):
+        self._filter = value
+
+    @property
+    def apodizer(self):
+        """Currently selected apodizer name"""
+        return self._apodizer
+    @apodizer.setter
+    def apodizer(self, value):
+        value = value.upper() # force to uppercase
+        if value not in self.apodizer_list:
+            raise ValueError("Instrument {0} doesn't have a apodizer called {1}.".format(self.name, value))
+        self._apodizer = value
+
+    @property
+    def fpm(self):
+        """Currently selected FPM name"""
+        return self._fpm
+    @fpm.setter
+    def fpm(self, value):
+        value = value.upper() # force to uppercase
+        if value not in self.fpm_list:
+            raise ValueError("Instrument {0} doesn't have a FPM called {1}.".format(self.name, value))
+        self._fpm = value
+
+    @property
+    def lyotstop(self):
+        """Currently selected Lyot stop name"""
+        return self._lyotstop
+    @lyotstop.setter
+    def lyotstop(self, value):
+        # preserve case for this one since we're used to that with the lyot mask names
+        if value not in self.lyotstop_list:
+            raise ValueError("Instrument {0} doesn't have a Lyot mask called {1}.".format(self.name, value))
+        self._lyotstop = value
+
+    @property
+    def mode_list(self):
+        """Available Observation Modes"""
+        keys = self._mode_table.keys()
+        keys = sorted(keys)
+        return keys
+
+    # mode works differently since it's a meta-property that affects the other ones:
+    @property
+    def mode(self):
+        """Currently selected mode name"""
+        for modename, settings in self._mode_table.items():
+            if (self.camera==settings[0].upper() and self.filter==settings[1].upper() and 
+                self.apodizer==settings[2].upper() and self.fpm==settings[3].upper() and 
+                self.lyotstop==settings[4]):
+                return modename
+        return 'Custom'
+    @mode.setter
+    def mode(self, value):
+        if value not in self.mode_list:
+            raise ValueError("Instrument {0} doesn't have a mode called {1}.".format(self.name, value))
+        settings = self._mode_table[value]
+        self.camera=settings[0]
+        self.filter=settings[1]
+        self.apodizer=settings[2]
+        self.fpm=settings[3]
+        self.lyotstop=settings[4]
+        _log.info('Set the following optical configuration:')
+        _log.info('camera = {0}, filter = {1}, apodizer = {2}, fpm = {3}, lyotstop = {4}'.format\
+                  (self.camera, self.filter, self.apodizer, self.fpm, self.lyotstop))
+
+        if self.mode is 'DISKSPC_F721':
+            self._fpmres = 3
+            self._owa = 20.
+        else: # for now, default to CHARSPC
+            self._fpmres = 4
+            self._owa = 9.
+        self._Mfpm = int(np.ceil(self._fpmres*self._owa))
+
+        if self.apodizer == 'DISKSPC':
+            self._apodizer_fname = \
```

Comment:
This part of the code should be in `apodizer.setter`, I think? Right now it will only get executed if the user sets the `mode`, but not if they set the `apodizer` property directly. Similar logic applies related to the lyot stop etc.

If your intent is to let users _only_ adjust these via setting the `mode`, then we could disable the setters for the other attributes, but that would be less consistent with how the other SIs work. I think it would be cleaner to have the `_apodizer_fname` and `_fpm_fname` etc all set inside the various attribute setter functions. Does that make sense?